### PR TITLE
Fix video stream demo image scaling, shaking and color shift

### LINKED ISSUE #39: Stream video example fails with the Framegrabber bouncing ball
When running the stream video example and streaming the Framegrabber "bouncing" ball, the bouncing ball does not appear at all.
There is another issue related to this problem: the FRamegrabber "line" vanishes for a few seconds when reaching the bottom of the canvas.

> ## WIP: Render the received samples in an HTML canvas as done in the `yarp.js` example https://github.com/robotology/yarp.js/blob/master/examples/stream_video/index.html
> 
> - I first tried to do it by directly calling `video_stream_element.getContext('2d').drawImage(canvasImg,0,0)` from https://github.com/ami-iit/yarp-openmct/blob/6a7258b90016ef93d1707bbe350c61c13478da5d/openmctStaticServer/realtime-telemetry-plugin.js#L5 . This fails as the canvas is intermittently updated.
> - I then tried skipping all Open-MCT plugins installations, and instead, just displaying the canvas.
>    => Scrolling line [´mode --line´]: unlike what we observe on the YARP viewer, when the line reaches the bottom of the canvas, it doesn´t instantly appear at the top. Instead, it remains invisible for a while, then appears at the top and scrolls faster than the YARP viewer.
>    => Boucing ball [´mode --ball´]: doesn´t appear at all.
> - For checking if there was an occasional issue with my setup, I went back to the good old [`yarp.js` stream video example](https://github.com/robotology/yarp.js/blob/master/examples/stream_video/index.html). It also fails. Checking the setup...
> 

_Originally posted by @nunoguedelha in https://github.com/ami-iit/yarp-openmct/issues/104#issuecomment-1083065846_

#### Fix image scaling
We set the destination image size to the canvas size. This way the image will be resized and properly fit the canvas at all times.
#### Fix image shaking
- The image was trembling due to the display of the HTML element 'write-time'. This element is the image frame refresh interval oscillating between ~9 and ~11 ms, which results in a element length toggling.
- As the image has a display 'inline-block' style, it flickers along with the 'write-time' length.
- We just use a line break to avoid this effect.

Fixed in #40 .

### PR DESCRIPTION
**Pull request by nunoguedelha**
Fixes #39 .

### CODE REVIEW

**Comment by nunoguedelha** [comment]
CC @S-Dafarra @riccardogrieco

**Comment by nunoguedelha** [comment]
I'm still finishing an approach based on @s-dafarra's suggestion, but it's already worth reviewing the implementation so far already discussed in the call.

**Review by traversaro — APPROVED** [review]
Thanks, seems ok. Not sure if you want to check if we can do just with YARP but anyhow we can always merge this and improve it later.

File: YarpJS/src/YarpJS_Image.cpp [hunk 1]
```diff
@@ -36,6 +49,17 @@ void YarpJS_Image::compress(int compression_quality)
...
 
+    int imgPixelCode = this->getYarpObj()->getPixelCode();
+    if (imgPixelCode != VOCAB_PIXEL_BGR
+        && imgPixelCode != VOCAB_PIXEL_BGRA) {
+        auto conversionCodePtr = yarpCode2colorConvCode.find(imgPixelCode);
+        if (conversionCodePtr != yarpCode2colorConvCode.end()) {
+            cv::cvtColor(internalImage, internalImage, conversionCodePtr->second, 0);
+        } else {
+            throw std::out_of_range("YarpJS_Image::compress(int compression_quality): pixel code not supported");
+        }
```
Comment (RiccardoGrieco) [inline YarpJS/src/YarpJS_Image.cpp:60]: I think that in order to slightly improve the readability, the else branch can be removed:

```suggestion
        if (conversionCodePtr == yarpCode2colorConvCode.end()) {
            throw std::out_of_range("YarpJS_Image::compress(int compression_quality): pixel code not supported");    
        } 
        cv::cvtColor(internalImage, internalImage, conversionCodePtr->second, 0);
```

Also I would avoid throwing `std::out_of_range` as it is commonly related to bad access to container objects. I think [`std::logic_error`](https://en.cppreference.com/w/cpp/error/logic_error) or [`std::domain_error`](https://en.cppreference.com/w/cpp/error/domain_error) could be a better fit for this case.

Comment (nunoguedelha) [reply YarpJS/src/YarpJS_Image.cpp:60]: Regarding the removal of `else`, absolutely.
Regarding the exception type, I kind of disagree:
- `std::logic_error` seems too generic to me, all the other types being derived from this one, so I really wouldn't pick this one.
- `std::domain_error` is suitable for an argument being out of range for a given operation (e.g. [math_errhandling](https://en.cppreference.com/w/cpp/numeric/math/math_errhandling)).
- **`std::out_of_range`** is actually the default exception used by `std::map` when you try to access a non existing key and that is why I picked it.

As you mentioned yourself, `std::out_of_range` is suited for bad access to containers. More precisely, it indicates an "attempt to access elements out of defined range." (refer to documentation). Well, at a lower level that's the case here. Then the custom message that follows brings some clarification.

**Comment by nunoguedelha** [comment]
> Thanks, seems ok. Not sure if you want to check if we can do just with YARP but anyhow we can always merge this and improve it later.

Thanks @traversaro , I will submit a separate PR for the improvement, as you suggested, that way it shall be easier to discuss around it and eventually discard it.

File: YarpJS/src/YarpJS_Image.cpp (hunk 1 above)
Comment (nunoguedelha) [reply YarpJS/src/YarpJS_Image.cpp]: As per what we agreed in our live discussion, I'm keeping your suggestion on the `<if/else>`, but keeping `std::out_of_range`.

**Pull request merged**